src/bloqade/gemini/analysis/measurement_validation/impls.py

```diff
@@ -0,0 +1,129 @@
+from kirin import ir, interp as _interp
+from kirin.analysis import ForwardFrame
+from kirin.dialects import func
+
+from bloqade import qubit, gemini
+from bloqade.analysis.address.lattice import AddressReg, AddressQubit
+from bloqade.analysis.measure_id.lattice import MeasureIdTuple
+
+from .analysis import _GeminiTerminalMeasurementValidationAnalysis
+
+
+@qubit.dialect.register(key="gemini.validate.terminal_measurement")
+class __QubitGeminiMeasurementValidation(_interp.MethodTable):
+
+    # This is a non-logical measurement, can safely flag as invalid
+    @_interp.impl(qubit.stmts.Measure)
+    def measure(
+        self,
+        interp: _GeminiTerminalMeasurementValidationAnalysis,
+        frame: ForwardFrame,
+        stmt: qubit.stmts.Measure,
+    ):
+
+        interp.add_validation_error(
+            stmt,
+            ir.ValidationError(
+                stmt,
+                "Non-terminal measurements are not allowed in Gemini programs!",
+            ),
+        )
+
+        return (interp.lattice.bottom(),)
+
+
+@gemini.logical.dialect.register(key="gemini.validate.terminal_measurement")
+class __GeminiLogicalMeasurementValidation(_interp.MethodTable):
+
+    # This is a logical terminal measurement, which is allowed
+    # but we impose the following restrictions:
+    # 1. All qubits spawned MUST be consumed
+    @_interp.impl(gemini.logical.stmts.TerminalLogicalMeasurement)
+    def terminal_measure(
+        self,
+        interp: _GeminiTerminalMeasurementValidationAnalysis,
+        frame: ForwardFrame,
+        stmt: gemini.logical.stmts.TerminalLogicalMeasurement,
+    ):
+
+        # should only be one MeasureIDTuple
+        interp.num_terminal_measurements += 1
+
+        if interp.num_terminal_measurements > 1:
+            interp.add_validation_error(
+                stmt,
+                ir.ValidationError(
+                    stmt,
+                    "Multiple terminal measurements are not allowed in Gemini logical programs!",
+                ),
+            )
+            return (interp.lattice.bottom(),)
+
+        # If we confirm there isn't a duplicate terminal measurement,
+        # now we need to check that for all the qubits that were spawned,
+        # they are all consumed by this measurement
+        address_analysis_results = interp.address_analysis_results
+        measurement_analysis_results = interp.measurement_analysis_results
+
+        # should just be one MeasureIDTuple
+        measure_lattice_element = measurement_analysis_results.get_values(stmt.results)[
+            0
+        ]
+
+        # Figure out the total number of qubits spawned, keeping in mind that if a user
+        # "shuffles" the qubits (puts them in a new container, splits one off from a container type, etc.)
+        # it should be accounted for. This would be much cleaner if there was a way to propagate the
+        # final qubit count saved in the actual interpreter for address analysis...
```

Comment:
I don't understand this comment: you're running the address analysis, right? Why can't you get the `AddressAnalysis.qubit_count` value at some point?

Reply:
I think I talked about this with Phil because I thought of this too, that unique qubit counter is attached to the analysis itself, not the frame - which is the only thing I have access to in validation. 

I don't think there's a nice way to affix that to the final frame? Could be wrong though

Reply:
Just putting this down for posterity, @david-pl pointed out I can just manually run the address analysis interpreter in the validation pass and route the info I need instead of relying on the ValidationPass auto-run infrastructure (: